libraries/adaptive-expressions/src/functionUtils.internal.ts

```diff
@@ -298,48 +298,84 @@ export class InternalFunctionUtils {
         }
 
         if (!error) {
-            const iteratorName = (expression.children[1].children[0] as Constant).value as string;
-            let arr = [];
-            if (Array.isArray(instance)) {
-                arr = instance;
-            } else if (typeof instance === 'object') {
-                Object.keys(instance).forEach((u): number => arr.push({ key: u, value: instance[u] }));
+            const list = InternalFunctionUtils.convertToList(instance);
+            if (!list) {
+                error = `${expression.children[0]} is not a collection or structure object to run Foreach`;
             } else {
-                error = `${expression.children[0]} is not a collection or structure object to run foreach`;
-            }
-
-            if (!error) {
-                const stackedMemory = StackedMemory.wrap(state);
                 result = [];
-                for (const item of arr) {
-                    const local: Map<string, any> = new Map<string, any>([[iteratorName, item]]);
-
-                    stackedMemory.push(SimpleObjectMemory.wrap(local));
-                    const { value: r, error: e } = expression.children[2].tryEvaluate(stackedMemory, options);
-                    stackedMemory.pop();
-                    if (e !== undefined) {
-                        return { value: undefined, error: e };
+                InternalFunctionUtils.lambdaEvaluator(expression, state, options, list, (currentItem, r, e) => {
+                    if (e) {
+                        error = e;
+                        return true;
+                    } else {
+                        result.push(r);
+                        return false;
                     }
-                    result.push(r);
-                }
+                });
             }
         }
 
         return { value: result, error };
     }
 
+    /**
+     * Lambda evaluator.
+     * @param expression expression.
+     * @param state memory state.
+     * @param options options.
+     * @param list item list.
+     * @param callback call back. return the should break flag.
+     */
+    public static lambdaEvaluator(expression: Expression, state: MemoryInterface, options: Options, list: unknown[], callback: (currentItem: unknown, result: unknown, error: string) => boolean): void{
```

Comment:
It's fairly simple to plumb through a type parameter that defaults to unknown that can provide some utility.

```typescript
public static lambdEvaluator<T = unknown, U = unknown>(expression: Expression, state: MemoryInterface, options: Options, list: T[], callback: (currentItem: T, result: U, error: string) => boolean {
  ...
}
```

The compiler can then infer types from usage which provides a better developer experience.